src/packages/piral-oidc/src/setup.ts

```diff
@@ -15,41 +22,57 @@ export function setupOidcClient(config: OidcConfig): OidcClient {
     responseType,
     scopes,
     restrict = false,
+    appUri,
+    logLevel,
   } = config;
+
   const userManager = new UserManager({
     authority: identityProviderUri,
     redirect_uri: redirectUri,
     silent_redirect_uri: redirectUri,
+    popup_redirect_uri: redirectUri,
     post_logout_redirect_uri: postLogoutRedirectUri,
     client_id: clientId,
     client_secret: clientSecret,
     response_type: responseType,
     scope: scopes?.join(' '),
   });
 
-  if (process.env.NODE_ENV === 'development') {
+  if (logLevel !== undefined) {
+    Log.logger = console;
+    Log.level = logLevel;
+  } else if (process.env.NODE_ENV === 'development') {
     Log.logger = console;
     Log.level = Log.DEBUG;
   }
 
-  userManager.signinRedirectCallback();
-  userManager.signinSilentCallback();
-  userManager.signoutRedirectCallback();
+  if (window.location.pathname === new URL(userManager.settings.post_logout_redirect_uri).pathname) {
+    if (window === window.top) {
+      userManager.signoutRedirectCallback();
+    } else {
+      userManager.signoutPopupCallback();
+    }
+  }
 
   const retrieveToken = () => {
     return new Promise<string>((res, rej) => {
-      userManager.getUser().then(
-        (user) => {
+      userManager
+        .getUser()
+        .then((user) => {
           if (!user) {
-            rej('Not logged in. Please call `login()` to retrieve a token.');
+            rej(notLoggedInMessage);
```

Comment:
Hm. Regarding these things: How should they be used? Because these messages are rather verbose as if they are intended for the end-user. However, since they are hardcoded and in English I don't see them as end-user messages.

I would prefer an error being put in here, which has a `type` and `message`. For the message we can discuss if these should be different depending on `NODE_ENV` (e.g., React has just a single code for production, but quite verbose / explicit ones in development).

The advantage of having a kind of error type (e.g., "unauthorized") is that the message can be changed in the future, and still code can react properly to the rejection scenarios.

Reply:
```typescript
const enum ErrorTypes {
    notAuthorized = 'notAuthorized',
    silentRenewFailed = 'silentRenewFailed',
}

const englishMessages = {
    [ErrorTypes.notAuthorized]: 'Not logged in. Please call `login()` to retrieve a token.',
    [ErrorTypes.silentRenewFailed]: 'Silent renew failed to retrieve access token.'
};

const getMessage = (code: ErrorTypes) => {
    return englishMessages[code];
}
```

Is this what you were thinking of? Throwing would be something like `new OidcError(ErrorTypes.notAuthorized);`

The interface would be something like
```typescript
interface IOidcError extends Error {
  type: ErrorTypes
}
```

Reply:
If we treat the messages as developer-related messages (which is now enabled by the error types) I would just call them "errorMessages". If somebody wants translations they should just use the error type and localize based on these keys.

Regarding the rest - did not think about it so far... I like the proposal and would go with it!

Reply:
I see these as developer messages as well, not end user messages. The end user may not even get an error in many implementations, because an error means they will get logged out, and the error logged to a logserver somewhere.

Reply:
Updated to use OidcErrorType and OidcError